mediagrains/utils/h264_grain_wrapper.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2021 British Broadcasting Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""\
+Support for reading H.264 essence data and wrapping it in Grains.
+This can be used for tasks like piping the output of ffmpeg (or similar) into a GSF file.
+"""
+import typing
+import copy
+from math import ceil
+
+from mediatimestamp.immutable import TimeRange
+
+from ..grain import GRAIN, CODEDVIDEOGRAIN
+from .h264_parser import H264Parser, FrameInfo
+
+
+class H264GrainWrapper(object):
+    """Raw input and wrap it in Grains"""
+    def __init__(
+        self,
+        template_grain: GRAIN,
+        input_data: typing.IO[bytes]
+    ):
+        """Set up the wrapper and the Grains that will be generated
+
+        :param template_grain: A coded video Grain to use as the template for wrapping the Grains read from the input source.
+                               origin_timestamp should be set.
+        :param input_data: An object to read video data from
+        """
+        assert(isinstance(template_grain, CODEDVIDEOGRAIN))
+        self.template_grain = copy.deepcopy(template_grain)  # make a copy as the template defaults will be updated
+        self.input_data = input_data
+        self.input_data_buffer = bytearray()
+
+    def _frames(self) -> typing.Iterator[typing.Tuple[bytes, typing.List[int], typing.Optional[FrameInfo]]]:
+        """Generator that yields H.264 frame bytes read from the input given
+
+        :yields: H.264 frame bytes objects read from the raw input supplied
+        """
+        h264_parser = H264Parser()
+
+        end_of_input_data = False
+        prev_frame_size = None
+        while not end_of_input_data:
+            # Parse the frame. The frame size is determined by detecting the start of the next frame
+            start_of_new_frame = True
+            while True:
+                # Estimate how much to read based on the previous frame size and default to 8K
+                read_size = 0
+                if start_of_new_frame:
+                    if prev_frame_size is not None:
+                        if prev_frame_size > len(self.input_data_buffer):
+                            read_size = prev_frame_size - len(self.input_data_buffer)
+                            read_size = ceil(read_size / 8192) * 8192
```

Comment:
The number 8192 is dotted around quite a bit here. I'm guessing it relates to the 8K but it isn't clear. Is it possible to put it in a named const?

Reply:
Will do. Yes, it's 8K, an arbitrary read block size.